sympy/concrete/expr_with_intlimits.py

```diff
@@ -283,3 +284,57 @@ def reorder_limit(expr, x, y):
             return type(expr)(expr.function, *limits)
         else:
             raise ReorderError(expr, "could not interchange the two limits specified")
+
+    @property
+    def has_empty_sequence(self):
+        """
+        Returns True if the Sum or Product is computed for an empty sequence.
+
+        Examples
+        ========
+
+        >>> from sympy import Sum, Product, Symbol
+        >>> m = Symbol('m')
+        >>> Sum(m, (m, 1, 0)).has_empty_sequence
+        True
+
+        >>> Sum(m, (m, 1, 1)).has_empty_sequence
+        False
+
+        >>> M = Symbol('M', integer=True, positive=True)
+        >>> Product(m, (m, 1, M)).has_empty_sequence
+        False
+
+        >>> Product(m, (m, 2, M)).has_empty_sequence
+
+        >>> Product(m, (m, M + 1, M)).has_empty_sequence
+        True
+
+        >>> N = Symbol('N', integer=True, positive=True)
+        >>> Sum(m, (m, N, M)).has_empty_sequence
+
+        >>> N = Symbol('N', integer=True, negative=True)
+        >>> Sum(m, (m, N, M)).has_empty_sequence
+        False
+
+        See Also
+        ========
+
+        has_reversed_limits
+        has_finite_limits
+
+        """
+        ret_None = False
+        for lim in self.limits:
+            dif = lim[1] - lim[2]
+            eq = Eq(dif, 1)
+            if eq == True:
+                return True
```

Comment:
How is Sum defined for non-integer limits like `Sum(f(n), (n, 1, S(3)/2))`?

Reply:
No idea... I still haven't understood if it is an actual summation using those properties or if it simply applies the closed form equations. 

Here is a good set of examples which I think shows that it is a combination of approaches:

```
from sympy import *
m, n, x = symbols('m n x')
summation(sin(x), (x, S.Half, S.Half+2))
summation(sin(x), (x, S.Half, 3))
summation(x, (x, S.Half, 3))
summation(x, (x, S.Half, n))
summation(sin(x), (x, m, n))
```

Earlier on today, I realized that 
```
summation(x, (x, S.Half, 3+I))
```
gives a numerical result...

(Will get back with more comments.)

Reply:
I think that the procedure is as follows:

If the difference between the upper and lower bound is a real integer it will try just replace the expression if there are less than 100 terms. (For reversed limits, it will flip it back first.)

If there is a symbolic solution is will just use that, without considering if the bounds are real or not and integer or not.

Which corresponds to the examples and the my rough understanding of the code.

Not sure if this is a very good approach though. Especially when it comes to allowing complex limits. As a consequence, conjugates are (currently) not propagated to the function. So at least it is consistent with the behaviour. Ping @asmeurer @smichr @jksuom : Do you have any opinions on if summations (and products) should be able to accept complex and/or non-integer arguments? I can see the point of being able to sum over arbitrary sets, but not really in the Sum_from^to-format.

Reply:
I would only use integer limits. If something else is needed, sum over `a, a + 1, a + 2, ..., a + n`, for example, then I would rewrite the index as `a + i`.

Reply:
Thanks @jksoum ! It will be a breaking change though if we force all limits to be integers:

https://github.com/sympy/sympy/blob/bfc42a2fc79deeec769761a2af6cadb1e621300e/sympy/concrete/tests/test_sums_products.py#L322-L327

But in my opinion it is worth it.

Reply:
After playing around, I think one reason to allow non-integer limits is that they can happen after change_index-rewrites.

For example, trying to solve:
`Sum(1/(x**2+2*sqrt(2)*x + 2), (x, K, M))`

One way here is to instead solve
`Sum(1/(x**2+2*sqrt(2)*x + 2), (x, K, M)).change_index(x, x+sqrt(2)).simplify(doit=False)`

this doesn't work due to an error in try_lerchphi, but 
`Sum(1/(x**2+2*sqrt(2)*x + 2), (x, K, M)).change_index(x, x+sqrt(2)).simplify(doit=False).subs({M + sqrt(2) : M, K + sqrt(2) : K}).doit().subs({M : M + sqrt(2), K : K + sqrt(2)})`
does (which numerically give the same for all M and K I've tested). Now, there is no actual summation with non-integer limits involved, but ideally, `doit()` should be able to do these transformations automatically (but then, I guess that it can also do the subs so maybe not a big issue anyway...).

Reply:
I think that non-integer limits can be okay as in something like:
```julia
In [3]: Sum(f(n), (n, -S(1)/2, S(1)/2)).doit()                                                                                    
Out[3]: f(-1/2) + f(1/2)
```
We should however require that the difference between the upper and lower limits be an integer so this is not okay:
```julia
In [4]: Sum(1, (n, 0, S(3)/2)).doit()                                                                                             
Out[4]: 5/2
```
The Sum should be required to have a nonnegative integer number of terms so if the summand is 1 then the result must always be an integer.

I think that we should also disallow "reversed" limits as in
```julia
In [7]: Sum(1, (n, 0, -2)).doit()                                                                                                 
Out[7]: -1
```
I think that if the upper limit is lower than the lower limit then the result should just be zero.

I propose the following equivalence:
```
Sum(f(n), (n, a, b)) = Sum(f(a+m), (m, 0, b-a)) = sum(f(a+m) for m in range(b-a+1))
```
This gives:
```julia
In [8]: def sumrange(expr, sym, a, b): 
   ...:     return sum(expr.subs(sym, a+m) for m in range(b-a+1)) 
   ...: 

In [13]: sumrange(f(n), n, 0, 3)                                                                                                  
Out[13]: f(0) + f(1) + f(2) + f(3)

In [14]: sumrange(f(n), n, S(1)/2, S(3)/2)                                                                                        
Out[14]: f(1/2) + f(3/2)

In [15]: sumrange(f(n), n, 0, -3)                                                                                                 
Out[15]: 0

In [16]: sumrange(f(n), n, 0, S(1)/2)                                                                                             
---------------------------------------------------------------------------
TypeError
```
This would mean that some general results should come back as Piecewise e.g.:
```julia
In [18]: Sum(1, (n, x, y)).doit()                                                                                                 
Out[18]: -x + y + 1

In [19]: Piecewise((1+y-x, x<=1+y))   # Correct answer
Out[19]: {-x + y + 1  for x ≤ y + 1
```
I think that it is okay to assume with symbolic limits that b-a is an integer if the assumptions don't justify it but if b-a is definitely not an integer then an error should be raised.

The docstring for Sum describes something different from what I suggest above but also doesn't seem to be consistent with current behaviour of actual Sum.